### PR TITLE
Set dbname and username when MyProcPort is not initialized when setup…

### PR DESCRIPTION
**Pull request by zhangh43**
… connection.

InitPostgres() in GPDB needs to recover DTM, and will setup connection
from QD to QE by calling cdbconn_doConnectStart(), cdbconn_doConnectStart()
uses MyProcPort to get username and dbname currently. But for background
worker, MyProcPort is not set. Set MyProcPort before InitPostgres() is not
appropriate since catalog is not initialized. We could use MyDatabaseId and
AuthenticatedUserId instead, which are initialized in InitPostgres() for
background worker.

### CODE REVIEW

**Comment by hlinnaka** [comment]
I don't understand this...

Is this fixing an existing bug, or?

Why does InitPostgres need to do DTM recovery? Where exactly does that happen? Does it need to do it in background workers?

In InitPostgres, at the point where MyDatabaseId is set, we're already accessing catalogs. So "Set MyProcPort before InitPostgres() is not appropriate since catalog is not initialized" doesn't seem right to me. (I'm not sure if filling in MyProcPort earlier would be a good idea, even if that wasn't a problem, though)

**Comment by zhangh43** [comment]
@hlinnaka 
InitTM() is called by cdb_setup() for each QD process, while cdb_setup() which is called in InitPostgres, Below are the condition to call cdb_setup(). 
```
InitPostgres_>
{
    if (!bootstrap && IsUnderPostmaster)
    {
		cdb_setup();
		on_proc_exit( cdb_cleanup, 0 );
    }
}
```

InitTM() will do TM recovery, which will connects to all QE and resolve all in-doubt txn.
Background worker will also run as QD to dispatch queries, so we should treat them the same as normal QD. 

"In InitPostgres, at the point where MyDatabaseId is set, we're already accessing catalogs. " Yes, Set MyProcPort inside InitPostgres() for background worker is OK but not before InitPostgres().  This solution is gp-specific. In postgresql, background worker doesn't initialized the MyProcPort struct inside InitPostgres(). It's also redundant, we could get username and database name in cdbconn_doConnectStart() in other ways listed in PR.

**Comment by zhangh43** [comment]
"Is this fixing an existing bug, or?" It's a bug, currently run background worker will coredump in cdbconn_doConnectStart() if you don't set MyProcPort explicitly.

**Comment by zhangh43** [comment]
@asimrp Background worker need to do DTM init, because background worker could also dispatch a query to all the segments just like normal QD.  DTM init will do DTM_Recovery, and background worker maybe the first postgres process to do this DTM recovery work when restarting the database.

Step to repeat the bug
```
cd contrib/worker_spi/
make install
gpconfig -c shared_preload_libraries -v "worker_spi"
# core generated when restart database.
gpstop -arf
```

core stack
```
#0  0x0000000000ad136e in cdbconn_doConnectStart (segdbDesc=0x2ffbd30,
    gpqeid=0x7fff94adaa30 "1;595411598304556;true;0;2",
    options=0x2ff5860 " -c gp_qd_hostname=10.146.0.5 -c gp_qd_port=15432 -c client_min_messages=notice -c commit_delay=0 -c commit_siblings=5 -c coredump_on_memerror=false -c DateStyle=ISO,\\ MDY -c dml_ignore_target_partiti"...) at cdbconn.c:348
#1  0x0000000000ad94c0 in cdbgang_createGang_async (segments=0x2f79980,
    segmentType=SEGMENTTYPE_EXPLICT_WRITER) at cdbgang_async.c:136
#2  0x0000000000ad7f35 in cdbgang_createGang (segments=0x2f79980,
    segmentType=SEGMENTTYPE_EXPLICT_WRITER) at cdbgang.c:93
#3  0x0000000000ad801f in AllocateGang (ds=0x2ff38d0, type=GANGTYPE_PRIMARY_WRITER,
    segments=0x2f79980) at cdbgang.c:130
#4  0x0000000000ad6348 in cdbdisp_dispatchCommandInternal (pQueryParms=0x2f78f10, flags=0,
    cdb_pgresults=0x7fff94adadb0) at cdbdisp_query.c:397
#5  0x0000000000ad620c in CdbDispatchCommand (
    strCommand=0xd27a08 "select gid from pg_prepared_xacts", flags=0, cdb_pgresults=0x7fff94adadb0)
    at cdbdisp_query.c:335
#6  0x0000000000b230c8 in gatherRMInDoubtTransactions () at cdbtm.c:2240
#7  0x0000000000b22f9f in recoverInDoubtTransactions () at cdbtm.c:2164
#8  0x0000000000b22e27 in recoverTM () at cdbtm.c:2114
#9  0x0000000000b210cb in initTM_recover_as_needed () at cdbtm.c:1148
#10 0x0000000000b21457 in initTM () at cdbtm.c:1314
#11 0x0000000000b26529 in cdb_setup () at cdbutil.c:885
#12 0x0000000000a5093a in InitPostgres (in_dbname=0x7fc7210dc0ec "postgres", dboid=0, username=0x0,
```

**Comment by asimrp** [comment]
Thank you @zhangh43 for the context.  Now I understand the issue better.

A broader discussion on how to leverage background workers in Greenplum is needed here.  E. g. a background worker needs to be aware of the postmaster's role (master/primary).  What value of `Gp_role` should a background worker assume?  Does it make sense for a background worker to be started on master as well as all primaries?  How about designating a background worker as master-only?

Secondly, DTM recovery is performed by the first backend process forked after postmaster startup.  This is incorrect.  I remember discussing this with @ashwinstar before.  Crash recovery should ideally be considered complete only after DTM is recovered.  This is a bigger code change but the right solution to the problem this PR is trying to solve.

The code change in this PR is harmless and seems to solve the problem.  However, if the `worker_spi` failure is not a burning issue, my suggestion is to open up the above mentioned discussions on the mailing list and close this PR.

File: src/backend/cdb/dispatcher/cdbconn.c [hunk 1]
```diff
@@ -345,16 +346,34 @@ cdbconn_doConnectStart(SegmentDatabaseDescriptor *segdbDesc,
...
-		nkeywords++;
+		values[nkeywords] = MyProcPort->database_name;	
 	}
+	else
+	{
+		/*
+		 * get database name from MyDatabaseId, which is initialized
+		 * in InitPostgres()
+		 */
+		values[nkeywords] = get_database_name(MyDatabaseId);
```
Comment (pengzhout) [inline src/backend/cdb/dispatcher/cdbconn.c]: maybe add a assertion here.

Comment (zhangh43) [reply src/backend/cdb/dispatcher/cdbconn.c]: assertion added

**Comment by kainwen** [comment]
@zhangh43 
Should the same change also apply to GDD?

**Pull request merged**